File: lectures/L07/conv_layer/python/conv_demo.py
"""Simple convolutional layer demo."""

import random

# pylint: disable = consider-using-enumerate

Matrix1d = list[float]
Matrix2d = list[Matrix1d]


def create_matrix2d(size: int) -> Matrix2d:
    """Create a two-dimensional matrix.

    Args:
        size: Size of the matrix.

    Returns:
        New matrix initialized with zeros.
    """
    matrix: Matrix2d = []
    for _ in range(size):
        row = [0.0] * size
        matrix.append(row)
    return matrix


def init_matrix2d(matrix: Matrix2d) -> None:
    """Initialize matrix with zeros.

    Args:
        matrix: The matrix to initialize.
    """
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            matrix[i][j] = 0.0


def relu_output(num: float) -> float:
    """Get ReLU output based on the given input.

    Args:
        num: ReLU input.

    Returns: The corresponding ReLU output.
    """
    return num if num > 0 else 0


def relu_delta(num: float) -> float:
    """Get ReLU delta based on the given input.

    Args:
        num: ReLU input.

    Returns: The corresponding ReLU delta.
    """
    return 1.0 if num > 0 else 0


class ConvLayer:
    """Convolutional layer implementation."""

    # pylint: disable = too-many-instance-attributes

    def __init__(self, input_size: int, kernel_size: int) -> None:

        if kernel_size == 0 or input_size == 0 or input_size < kernel_size:
            raise ValueError(
                "Cannot create convolutional layer: invalid input arguments!"
            )

        pad_offset = kernel_size // 2
        padded_size = input_size + 2 * pad_offset

        self.input_padded = create_matrix2d(padded_size)
        self.input_gradients_padded = create_matrix2d(padded_size)
        self.input_gradients = create_matrix2d(input_size)
        self.kernel = create_matrix2d(kernel_size)
        self.kernel_gradients = create_matrix2d(kernel_size)
        self.output = create_matrix2d(input_size)
        self.preactivation_output = create_matrix2d(input_size)
        self.bias = random.random()
        self.bias_gradient = 0.0

        for ki in range(len(self.kernel)):
            for kj in range(len(self.kernel)):
                self.kernel[ki][kj] = random.random()

    def feedforward(self, input_data: Matrix2d) -> bool:
        """Perform feedforward operation.

        Args:
            input_data: Input data.

        Returns:
            True on success, false on failure.
        """
        if len(input_data) != len(self.output):
            return False

        if not self._pad_input(input_data):
            return False

        # Perform convolution and apply the activation function.
        for i in range(len(self.output)):
            for j in range(len(self.output)):
                num = self.bias
                for ki in range(len(self.kernel)):
                    for kj in range(len(self.kernel)):
                        num += self.input_padded[i + ki][j + kj] * self.kernel[ki][kj]
                self.preactivation_output[i][j] = num
                self.output[i][j] = relu_output(num)
        return True

    def backpropagate(self, output_gradients: Matrix2d) -> bool:
        """Perform backpropagation.

        Args:
            output_gradients: Output gradients from the next layer.

        Returns:
            True on success, false on failure.
        """
        if len(output_gradients) != len(self.output):
            return False

        init_matrix2d(self.input_gradients_padded)
        init_matrix2d(self.input_gradients)
        init_matrix2d(self.kernel_gradients)
        self.bias_gradient = 0.0

        for i in range(len(self.output)):
            # Check if the output gradient matrix is square, return false if not.
            if len(output_gradients[i]) != len(output_gradients):
                return False
            for j in range(len(self.output)):
                delta = output_gradients[i][j] * relu_delta(
                    self.preactivation_output[i][j]
                )
                self.bias_gradient += delta
                for ki in range(len(self.kernel)):
                    for kj in range(len(self.kernel)):
                        self.input_gradients_padded[i + ki][j + kj] += (
                            self.kernel[ki][kj] * delta
                        )
                        self.kernel_gradients[ki][kj] += (
                            self.input_padded[i + ki][j + kj] * delta
                        )
        self._extract_input_gradients()
        return True

    def optimize(self, learning_rate: float) -> bool:
        """Perform optimization.

        Args:
            learning_rate: Learning rate used to adjust the trainable parameters.

        Returns:
            True on success, false on failure.
        """
        if learning_rate <= 0.0:
            return False

        self.bias += self.bias_gradient * learning_rate
        for ki in range(len(self.kernel)):
            for kj in range(len(self.kernel)):
                self.kernel[ki][kj] += self.kernel_gradients[ki][kj] * learning_rate
        return True

    def _pad_input(self, input_data: Matrix2d) -> bool:
        init_matrix2d(self.input_padded)
        offset = len(self.kernel) // 2

        # Copy the input data to the padded input matrix.
        for i in range(len(input_data)):
            # Check if the input matrix is square, return false if not.
            if len(input_data[i]) != len(input_data):
                return False
            for j in range(len(input_data)):
                self.input_padded[i + offset][j + offset] = input_data[i][j]
        return True

    def _extract_input_gradients(self) -> None:
        offset = len(self.kernel) // 2

        # Extract the input gradients from the corresponding padded matrix.
        for i in range(len(self.output)):
            for j in range(len(self.output)):
                self.input_gradients[i][j] = self.input_gradients_padded[i + offset][
                    j + offset
                ]


def main() -> None:
    """Create and demonstrate a simple convolutional layer."""

    # Example 4x4 input matrix (could represent an image or feature map).
    input_data = [
        [1, 1, 1, 1],
        [1, 0, 0, 1],
        [1, 0, 0, 1],
        [1, 1, 1, 1],
    ]

    # Example output gradients (same shape as output, used for backpropagation demo).
    output_gradients = [
        [1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1],
    ]

    # Create a convolutional layer: 4x4 input, 2x2 kernel.
    conv_layer = ConvLayer(4, 2)

    print("Convolution input_data (2D):")
    for row in input_data:
        print("  ", " ".join(f"{num:.1f}" for num in row))

    conv_layer.feedforward(input_data)
    print("\nConvolution output (2D):")
    for row in conv_layer.output:
        print("  ", " ".join(f"{num:.1f}" for num in row))

    print("\nConvolution output gradients (2D):")
    for row in output_gradients:
        print("  ", " ".join(f"{num:.1f}" for num in row))

    conv_layer.backpropagate(output_gradients)
    print("\nInput gradients after backpropagation (2D):")
    for row in conv_layer.input_gradients:
        print("  ", " ".join(f"{num:.1f}" for num in row))


# Invoke the main function if this is the startup script.
if __name__ == "__main__":
    main()

File: lectures/L07/conv_layer/cpp/conv_demo.cpp
/**
 * @file Simple convolutional layer demo.
 */
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <vector>

namespace
{
/** One-dimensional matrix. */
using Matrix1d = std::vector<double>;

/** Two-dimensional matrix. */
using Matrix2d = std::vector<Matrix1d>;

/**
 * @brief Initialize given matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 */
void initMatrix(Matrix2d& matrix) noexcept
{
    // Fill the matrix with zeros.
    for (auto& row : matrix)
    {
        for (auto& num : row)
        {
            num = 0.0;
        }
    }
}

/**
 * @brief Initialize matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 * @param[in] size The desired size of the matrix.
 */
void initMatrix(Matrix2d& matrix, const std::size_t size)
{
    // Resize the matrix if necessary, then fill the matrix with zeros.
    matrix.resize(size, Matrix1d(size));
    initMatrix(matrix);
}

/**
 * @brief Check whether given matrix is square.
 *
 * @param[in] matrix The matrix to check.
 *
 * @return True if the matrix is square, false otherwise.
 */
[[nodiscard]] bool isMatrixSquare(const Matrix2d& matrix) noexcept
{
    // Check the matrix row by row.
    for (const auto& row : matrix)
    {
        // Return false if the matrix isn't square.
        if (row.size() != matrix.size()) { return false; }
    }
    // Return true to indicate that the matrix is square.
    return true;
}

/**
 * @brief Print the contents of given matrix.
 *
 * @param[in] matrix The matrix to print.
 */
void printMatrix(const Matrix2d& matrix) noexcept
{
    // Print the matrix contents row by row.
    for (const auto& row : matrix)
    {
        const auto* lastNum{row.empty() ? nullptr : &row[row.size() - 1U]};
        std::printf("\t");

        // Separate each number in the row with a comma.
        for (const auto& num : row)
        {
            std::printf("%.1f", num);
            if (&num < lastNum) { std::printf(", "); }
        }
        std::printf("\n");
    }
    std::printf("\n");
}

/**
 * @brief Generate a random starting value between 0.0 and 1.0.
 *
 * @return Random floating-point value in the range [0.0, 1.0].
 */
double randomStartVal() noexcept
{
    constexpr auto max = static_cast<double>(RAND_MAX);
    return std::rand() / max;
}

/**
 * @brief ReLU activation function (output).
 *
 * @param[in] input Input value.
 *
 * @return Output after ReLU activation.
 */
constexpr double reluOutput(const double input) noexcept { return 0.0 < input ? input : 0.0; }

/**
 * @brief ReLU activation function (derivative).
 *
 * @param[in] input Input value.
 *
 * @return Derivative of ReLU at input.
 */
constexpr double reluDelta(const double input) noexcept { return 0.0 < input ? 1.0 : 0.0; }

namespace ml
{
/**
 * @brief Convolutional layer structure.
 *
 * @todo Implement this struct!
 */
struct ConvLayer final
{
    /**
     * @brief Constructor.
     *
     * @param[in] inputSize Input size. Must be greater than 0.
     * @param[in] kernelSize Kernel size. Must be greater than 0 and smaller than the input size.
     */

    /**
     * @todo Check the input arguments in the constructor:
     *
     *       If the input size is 0, the kernel size is 0, or the input size is smaller than
     *       the kernel size, terminate the program with error message
     *       "Cannot create convolutional layer: invalid input arguments!".
     */

    /**
     * @brief Perform feedforward operation.
     *
     * @param[in] input Matrix holding input data.
     *
     * @return True on success, false on failure.
     */

    /**
     * @brief Perform backpropagation.
     *
     * @param[in] outputGradients Matrix holding gradients from the next layer.
     *
     * @return True on success, false on failure.
     */

    /**
     * @brief Perform optimization.
     *
     * @param[in] learningRate Learning rate to use. Must be in range (0.0, 1.0].
     *
     * @return True on success, false on failure.
     */

    /** @todo Define member variables here! Keep them public and skip the 'my' prefix. */

private:
    /**
     * @brief Pad input with zeros.
     *
     * @param[in] input Input data.
     */

    /**
     * @brief Extract input gradients.
     */
};
} // namespace ml
} // namespace

/**
 * @brief Create and demonstrate a simple convolutional layer.
 *
 * @return 0 on success, -1 on failure.
 */
int main()
{
//! @todo Remove this header guard (and/or uncomment the compiler flags in the makefile) once the
//        implementation is finished.
#ifdef CONV_LAYER_IMPLEMENTED

    // Example 4x4 input matrix (could represent an image or feature map).
    const Matrix2d input{{1, 1, 1, 1}, {1, 0, 0, 1}, {1, 0, 0, 1}, {1, 1, 1, 1}};

    // Example output gradients (target output for demonstration).
    const Matrix2d outputGradients{{1, 1, 1, 1}, {1, 1, 1, 1}, {1, 1, 1, 1}, {1, 1, 1, 1}};

    // Initialize the random generator with the current time as seed.
    std::srand(std::time(nullptr));

    // Create a convolutional layer: 4x4 input, 2x2 kernel.
    constexpr std::size_t inputSize{4U};
    constexpr std::size_t kernelSize{2U};
    ml::ConvLayer convLayer{inputSize, kernelSize};

    // Show the input matrix.
    std::printf("Convolution input data (2D):\n");
    printMatrix(input);

    // Perform feedforward (convolution).
    convLayer.feedforward(input);
    std::printf("Convolution output (2D):\n");
    printMatrix(convLayer.output);

    // Show the output gradients.
    std::printf("Convolution output gradients (2D):\n");
    printMatrix(outputGradients);

    // Perform backpropagation.
    convLayer.backpropagate(outputGradients);
    std::printf("Input gradients after backpropagation (2D):\n");
    printMatrix(convLayer.inputGradients);
    return 0;

//! @todo Remove this header guard (and/or uncomment the compiler flags in the makefile) once the
//        implementation is finished.
#endif /** CONV_LAYER_IMPLEMENTED */
}
